File: Clients/AndroidQ/QAnd/src/com/qframework/core/ServerkoParse.java
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

package com.qframework.core;

import java.io.IOException;
import java.io.InputStream;
import java.util.NoSuchElementException;
import java.util.StringTokenizer;
import java.util.Vector;

import org.json.JSONObject;

import com.qframework.serverko.ServerkoConnect;

import android.util.Log;
import android.webkit.JsResult;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.webkit.WebChromeClient;
import android.os.Handler;
import android.os.Message;

public class ServerkoParse {

    private static final int EXECSCRIPT= 3; 
    private static final int LOADMODULE= 4; 
    private static final int LOADMODULE2= 5;

    private WebView mScriptView;
	private GameonApp mApp;
	private String			mUsername = "LUser";
	private Handler mMessageHandler;   
	private Vector<String>	mLoadedScripts = new Vector<String>();
	private ServerkoConnect 	mConnect;
   
	ServerkoParse(GameonApp theApp) {
	
		mApp = theApp;
		mScriptView = new WebView(theApp.context());
		WebSettings settings = mScriptView.getSettings();
		settings.setCacheMode(WebSettings.LOAD_NO_CACHE);
		settings.setNavDump(false);

		mScriptView.getSettings().setJavaScriptEnabled(true);
		
		//mScriptView.getSettings().setJavaScriptCanOpenWindowsAutomatically(true);
		mScriptView.setWebViewClient(new WebViewClient() {
			   public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
			     //Toast.makeText(activity, "Oh no! " + description, Toast.LENGTH_SHORT).show();
				   System.out.println(" Errror:" + description);
			   }
			 });		
		mScriptView.setWebChromeClient(new QWebChromeClient() );
		mMessageHandler = new Handler() { 

	        @Override 
	        public void handleMessage(Message msg) { 
	             switch (msg.what) { 
	             case LOADMODULE: 
	            	 execLoadModule((String)msg.obj);
	                 break;
	             case LOADMODULE2: 
	            	 execLoadModule2((String)msg.obj);
	                 break;	                 
	             case EXECSCRIPT: 
	            	 execScript((String)msg.obj);
	                 break;	                 	                 
	             default:
	                 super.handleMessage(msg); 

	             } 
	        } 
	   }; 	
		
	}

	
	public void execScript(String script) {
	
		//mScriptView.loadUrl("javascript:"  + "try{" + script + "} catch (error){ alert( \"error: \" + error.toString() );for (var i in error) console.error(i + ' = ' + error[i]);}");
		mScriptView.loadUrl("javascript:"  + script );
		
	}
	public void execScriptRaw(String script) {
		
		mScriptView.loadUrl( script );
		
	}

	public void execScriptEvent(String script, int counter) {
		//Log.d("model", "cnt = " + counter);
		if (script.startsWith("cmd?"))
		{
			String data = script.substring(4);
			//mScriptView.loadUrl("javascript:"  + "try{" + data + "} catch (error){ alert( \"error: \" + error.toString() );}for (var i in error) console.error(i + ' = ' + error[i]);");	
			mScriptView.loadUrl("javascript:"  + data );
		}else if (script.startsWith("include?"))
		{
			String data = script.substring(8);
			if (mLoadedScripts.indexOf(data) < 0)
			{
				mLoadedScripts.add(data);
				this.execScriptFromFile(data);
			}
		}else if (script.startsWith("load?"))
		{
			String data = script.substring(5);
			this.execScriptFromFile(data);
		}
			
	}

	  final class QWebChromeClient extends WebChromeClient {

	        @Override
	        // used for script log
	        public boolean onJsAlert(WebView view, String url, String message, JsResult result) {
	            Log.d("model", "logalert->" + message);
	            result.confirm();
	            return true;
	        }
	        public void onConsoleMessage (String message, int lineNumber, String sourceID) {
	            //Log.d(LOG_TAG, "error->" + message + " " + lineNumber + " " + sourceID);
	        }

	        public boolean onJsConfirm (WebView view, String url, String message, JsResult result)
	        {
	        	result.confirm();
	        	if (message.length() > 0)
	        	{
		        	//Log.d("model", "logconfirm2->" + message);
		        	JSONObject response  = MessageParse.parse2( message);
		        	if (response != null)
		        	{
		        		mApp.queueResponse(response);
		        	}
	        	}
	        	return true;
	        }
	    }
	public void execScriptFromFile(String fname) {
		try {
			InputStream is = mApp.context().getAssets().open("res/"+fname);
            int size = is.available();
            // Read the entire asset into a local byte buffer.
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();

            String script = new String(buffer);
			//mScriptView.loadUrl("javascript:" + "try{" + script + "} catch (error){ alert( \"error: \" + error.toString() );}for (var i in error) console.error(i + ' = ' + error[i]);");
            mScriptView.loadUrl("javascript:" + script );
			
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
	}

	public void execQScriptFromFile(String fname) {
		try {
			InputStream is = mApp.context().getAssets().open("qres/"+fname);
            int size = is.available();
            // Read the entire asset into a local byte buffer.
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();

            String script = new String(buffer);
            mScriptView.loadUrl("javascript:" +script);
			
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
	}
	
	public void sendUserData(String datastr) {
		execScript("Q.handlers.script_userData('"+mUsername+"', '" + datastr + "');");
	}	 


	public void loadFramework()
	{
		execQScriptFromFile("serverkobridge.js");
		execQScriptFromFile("room.js");
		execQScriptFromFile("layout.js");
		execQScriptFromFile("colors.js");    
		execQScriptFromFile("qframework.js");
	    
	}


	public void loadModule2(String resptype) {
		Message msg = new Message(); 
        msg.what = LOADMODULE2; 
        msg.obj = resptype;	            
        mMessageHandler.sendMessageDelayed(msg, 0);				
	}
	
	public void loadScript(final String resptype, int time) {
		
		loadModule2(resptype);
	}
	
	protected void loadModule(String resptype) {
		Message msg = new Message(); 
        msg.what = LOADMODULE; 
        msg.obj = resptype;	            
        mMessageHandler.sendMessageDelayed(msg, 0);		
	}


	protected void execScript(String respdata, int delay) {
		Message msg = new Message(); 
        msg.what = EXECSCRIPT; 
        msg.obj = respdata;	            
        mMessageHandler.sendMessageDelayed(msg, delay); 
		
	}

	protected void execLoadModule(String data)
	{
		if (mLoadedScripts.indexOf(data) < 0)
		{
			mLoadedScripts.add(data);
			this.execScriptFromFile(data);
		}
		 
	}
	
	protected void execLoadModule2(String data)
	{
		this.execScriptFromFile(data);		 
	}

	protected void connect(String serverip, String script) {
		if (mConnect == null)
		{
			mConnect = new ServerkoConnect();
		}
		
		mConnect.connect(serverip, script, this);
	}

	protected void disconnect() {
		if (mConnect != null)
		{
			mConnect.close();
		}
		
		
	}
	
	protected void join(String addr, String user, String script) {
		if (mConnect != null)
		{
			mConnect.join(addr, user, script);
			mUsername = user;
		}
	}

	protected void send(String data) {
		if (mConnect != null)
		{
			mConnect.send(data);
		}
		
	}

	public void onJSONData(String data) {
    	JSONObject response  = MessageParse.parse2( data);
    	if (response != null)
    	{
    		mApp.queueResponse(response);
    	}                	
		
	}

	public static int parseFloatArray(float[] array, String data)
	{
		int count = 0;
		float val = 0;
        StringTokenizer tok = new StringTokenizer(data, ",");
        while (tok.hasMoreTokens() && count < array.length)
        try {
            try {
                val= Float.parseFloat(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            array[count++] = val;
        } catch (NoSuchElementException e) {
        }
		return count;
	}

	public static int parseIntArray(int[] array, String data)
	{
		int count = 0;
		int val = 0;
        StringTokenizer tok = new StringTokenizer(data, ",");
        while (tok.hasMoreTokens() && count < array.length)
        try {
            try {
                val= Integer.parseInt(tok.nextToken());
            } catch (NumberFormatException e) {
            }
            array[count++] = val;
        } catch (NoSuchElementException e) {
        }
		
        return count;
	}

}

File: Clients/iPhoneQ/Q/QFramework/ServerkoParse.m
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

#import "ServerkoParse.h"
#import "GameonApp.h"
#import "AsyncSocket.h"
#import "JSON.h"

@implementation ServerkoParse



- (id)init
{
    self = [super init];
    
    if (self) {
        
        mJSServerko =  [[ServerkoJS alloc ] initWithDelegate:self];
        mActive = false;
        mAsyncSocket = nil;
        mCallbackScript = nil;
        mInConnect = false;
        
        mConnected = false;
        mInRoom  = false;  
        mWaitingForContent = false;
        mWaitingForJoin = false;
        mUsername = @"luser";
    }
    
    return self;
}



-(void) dealloc 
{
    [mJSServerko release];
    [mRoomId release];    
    [self disconnect];
    [super dealloc];
}


-(void) sendUserData:(NSString*) datastr onclick:(NSString*)click {
    if (!mActive)
    {
        return;
    }
    
    if ([click hasPrefix:@"js:"])
    {
        NSString* script = [click substringFromIndex:3];
        [mJSServerko execScript:script];        
    }else {
        NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
        [userdata appendFormat:@"Q.handlers.script_userData('%@', '%@');",mUsername,datastr];
        [mJSServerko execScript:userdata];
        
    }
}	 


- (void) execScript:(NSString*) script
{
    [mJSServerko execScript:script];
}

- (void)serverkoJS:(ServerkoJS*)serverkoJS onEvent:(NSMutableArray*)response {
    if (!mActive)
    {
        return;
    }    
    //[self output:response];
//    [mApp onData:response];
    [mApp queueResponses:response];
    //[ServerkoResponse releaseData:response];
    
}

- (void)serverkoJS:(ServerkoJS*)serverkoJS onLayout:(NSMutableArray*)response {
    if (!mActive)
    {
        return;
    }    
    //[self output:response];
//    [mApp onData:response];
    [mApp queueResponses:response];
    //[ServerkoResponse releaseData:response];
    
}

- (void)setApp:(GameonApp*) app
{
    mApp = app;
}


- (void)stopScript
{
    mActive = false;
    [NSObject cancelPreviousPerformRequestsWithTarget:mJSServerko ];
    [mJSServerko stopScript];
}


-(void) execScript:(int)time script:(NSString*)respdata
{

    double delay = (float)time * 0.001;
    
    if (delay < 1)
    {
        delay = 1;
    }
    
    [mJSServerko performSelector:@selector(execScript:) withObject:respdata afterDelay:delay];
    
    
    
}


-(void)loadScript:(NSString*)file delay:(int)scriptdelay
{
    double delay = (float)scriptdelay * 0.001;
    if (delay < 0.000001)
    {
        [mJSServerko performSelector:@selector(parse:) withObject:file];
        return;
    }
    
    [mJSServerko performSelector:@selector(parse:) withObject:file afterDelay:delay];
    
	
}


-(void)loadModule:(NSString*)file
{
//    [mJSServerko parse:file];
    [mJSServerko performSelector:@selector(parse:) withObject:file];
}

-(void)loadModule2:(NSString*)file
{
    [mJSServerko performSelector:@selector(loadScript:) withObject:file];
//    [mJSServerko mJSServerko:file];

}

-(void)initSocket
{
//	dispatch_queue_t mainQueue = dispatch_get_main_queue();	
	mAsyncSocket = [[AsyncSocket alloc] initWithDelegate:self];
    
}

- (void)onSocket:(AsyncSocket *)sock didConnectToHost:(NSString *)host port:(UInt16)port
{
    NSLog(@" connected ");
    mInConnect = false;
    if (mCallbackScript)
    {
        mConnected = true;
        NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
        [userdata appendFormat:@"%@(1, '%@:%d');",mCallbackScript,host,port];
        [mJSServerko execScript:userdata];
        [mAsyncSocket readDataToData:[AsyncSocket CRLFCRLFData] withTimeout:-1 tag:0];
        
    }
}

- (void) sendJoin:(int) data
{
    if (mCallbackScriptJoin)
    {
        NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
        [userdata appendFormat:@"%@(%d);",mCallbackScriptJoin,data];
        [mJSServerko execScript:userdata];
        
    }
        
}

- (void)onSocketDidDisconnect:(AsyncSocket *)sock
{
    mInConnect = false;
    if(mCallbackScript)
    {
        if (mConnected)
        {
            NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
            [userdata appendFormat:@"%@(2);",mCallbackScript];
            [mJSServerko execScript:userdata];
            
        }else {
            NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
            [userdata appendFormat:@"%@(0);",mCallbackScript];
            [mJSServerko execScript:userdata];
            
        }

        mConnected = false;
    }
    // start reading    
    
    
    NSLog(@" disconnected ");
}


-(void)connect:(NSString*)serverip callback:(NSString*)script
{
    if (mInConnect)
        return;
    
    if (mAsyncSocket == nil)
    {
        [self initSocket];
    }
    mCallbackScript = [[NSString alloc] initWithString:script];
    mInConnect = true;
    NSArray* tokens = [serverip componentsSeparatedByString:@":"];
    NSString* host = [tokens objectAtIndex:0];
    int port = [[tokens objectAtIndex:1] intValue];
    
	NSError *error = nil;
    if (![mAsyncSocket connectToHost:host onPort:port error:&error])
	{
		NSLog(@"Error connecting: %@", error);
        mInConnect = false;
        if (mCallbackScript)
        {
            NSMutableString* userdata = [[[NSMutableString alloc] init] autorelease];
            [userdata appendFormat:@"%@(0, '%@:%d');",mCallbackScript,host,port];
            [mJSServerko execScript:userdata];        
        }
	}
    
}

- (void)doSocketData:(NSString*)data
{
    if (mWaitingForContent)
    {
        [mJSServerko onJSONData:data];
        mWaitingForContent = false;
    }
    else {
        //analyze
        NSArray* tokens = [data componentsSeparatedByString:@"\r\n"];
        for (int a=0; a<[tokens count]; a++)
        {
            NSString* token = [tokens objectAtIndex:a];
            if ([token hasPrefix:@"HTTP/1.0 300"])
            {
                // multiple_choices 
                // user already exists
                [self sendJoin:-1];
                return;
            }
            else if ([token hasPrefix:@"HTTP/1.0 202"])
            {
                if (mWaitingForJoin)
                {
                    mWaitingForJoin = false;
                    [self sendJoin:1 ];
                }
            }            
            if ( [token hasPrefix:@"Content-length"] )
            {
                mWaitingForContent = true;
                return;
            }
        }
    }


}

- (void)onSocket:(AsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag
{
    NSString* strdata = [[NSString alloc] initWithBytes:[data bytes] length:[data length] encoding:NSUTF8StringEncoding];
    //NSLog(@"Received Data: %@", strdata);
    [self doSocketData:strdata];
    [strdata release];

    [mAsyncSocket readDataToData:[AsyncSocket CRLFCRLFData] withTimeout:-1 tag:0];
}


-(NSMutableString*) getJoinReq:(NSString*)room user:(NSString*)userid
{
    NSMutableString* strdata = [[NSMutableString alloc] initWithFormat:@"POST /10000/%@/%@ HTTP/1.1\r\n" , room, userid];    
    [strdata appendFormat:@"Host: www.100kas.com\r\n"];
    [strdata appendFormat:@"User-Agent: iosapp\r\n"];    
    [strdata appendFormat:@"\r\n"];
    return strdata;
}

-(NSMutableString*) getDataReq:(NSString*)data
{
    NSMutableString* strdata = [[NSMutableString alloc] initWithFormat:@"PUT /10000/%@/%@ HTTP/1.1\r\n" , mRoomId, mUsername];    
    [strdata appendFormat:@"Host: www.100kas.com\r\n"];
    [strdata appendFormat:@"User-Agent: iosapp\r\n"];        
    [strdata appendFormat:@"Data: %@\r\n", data];
    [strdata appendFormat:@"\r\n"];
    return strdata;
}

-(void)join:(NSString*)room user:(NSString*)userid callback:(NSString*)script
{

    mCallbackScriptJoin = [[NSString alloc] initWithString:script];
    NSLog(@" join room : %@ %@ %@" ,room, userid, script); 
    [mRoomId release];
    [mUsername release];
    
    mRoomId = [[NSString alloc] initWithString:room];
    mUsername = [[NSString alloc] initWithString:userid];    
    NSMutableString* req = [self getJoinReq:room user:userid]; 
    NSData *reqData = [req dataUsingEncoding:NSUTF8StringEncoding];
    mWaitingForJoin = true;

    [mAsyncSocket writeData:reqData withTimeout:1000 tag:0];
    [req release];
}

-(void)send:(NSString*)data
{
    
    NSMutableString* req = [self getDataReq:data ]; 
    NSData *reqData = [req dataUsingEncoding:NSUTF8StringEncoding];
    
    [mAsyncSocket writeData:reqData withTimeout:1000 tag:0];
    [req release];
}

-(void) disconnect
{
    if (mAsyncSocket == nil)
    {
        return;
    }
    [mAsyncSocket setDelegate:nil];
    [mAsyncSocket disconnect];
    [mAsyncSocket release];
    [mCallbackScript release];
    [mCallbackScriptJoin release];
    mCallbackScript = nil;
    mCallbackScriptJoin = nil;
    mAsyncSocket = nil;
    mInConnect = false;
    mConnected = false;
    mInRoom  = false;  
    mWaitingForContent = false;
    mWaitingForJoin = false;
    
}

+(int) parseFloatArray:(float*)array max:(int)max forData:(NSString*)data
{
	NSArray* tok = [data componentsSeparatedByString:@","];
	int count;
	for (count=0; count < [tok count] &&  count < max; count++)
	{
		array[count] = [[tok objectAtIndex:count ] floatValue];
	}
	
	return count;
}

+(int) parseIntArray:(int*)array max:(int)max forData:(NSString*) data
{
	NSArray* tok = [data componentsSeparatedByString:@","];
	int count;
	for (count=0; count < [tok count] &&  count < max; count++)
	{
		array[count] = [[tok objectAtIndex:count ] intValue];
	}
	
	return count;
}

-(void)start
{
    mActive = true;    
    [mJSServerko loadFramework];
    [mJSServerko runLoggers];

}
@end
